### PR TITLE
fuzzymatching

### PR DESCRIPTION
**Pull request by pavertomato**
[#4] + fuzzymatching in qido request

### CODE REVIEW

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java [hunk 1]
```diff
@@ -319,40 +319,32 @@ public static String getImageUrl(DicomStore store, String studyId) {
...
-        if (isNotBlank(query.getPhysicianName())) {
-            parameters.add("ReferringPhysicianName=" + urlEncode(query.getPhysicianName()));
-        }
-
-        if (parameters.isEmpty()) {
-            return allItems;
+        if (Objects.nonNull(query)) {
+            if (isNotBlank(query.getPatientName())) {
+                parameters.add("PatientName=" + urlEncode(query.getPatientName()));
+                parameters.add("fuzzymatching=true");
```
Comment (poojavenkatram) [inline src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: It's not necessary that whenever the patient name is matched it is a fuzzymatch - we should probably allow the user to choose whether to fuzzymatch or if they want exact matching.

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java [hunk 2]
```diff
@@ -319,40 +319,32 @@ public static String getImageUrl(DicomStore store, String studyId) {
...
-            );
-        }
 
-        if (isNotBlank(query.getPhysicianName())) {
-            parameters.add("ReferringPhysicianName=" + urlEncode(query.getPhysicianName()));
-        }
-
-        if (parameters.isEmpty()) {
-            return allItems;
+        if (Objects.nonNull(query)) {
```
Comment (poojavenkatram) [inline src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: I would make the null case the first step to reduce nesting and just return the default if the query was null.

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java [hunk 3]
```diff
@@ -319,40 +319,32 @@ public static String getImageUrl(DicomStore store, String studyId) {
...
+            if (isNotBlank(query.getPhysicianName())) {
+                parameters.add("ReferringPhysicianName=" + urlEncode(query.getPhysicianName()));
+            }
         } else {
-            return "?" + join(parameters, "&");
+            parameters.add("includefield=all");
         }
-    }
 
+        return "?" + join(parameters, "&");
```
Comment (poojavenkatram) [inline src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: Does join handle the case where parameters are empty skips adding any '&'?

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 2 above)
Comment (pavertomato) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: are you sure it's good idea to have two return statement in method? it's splits simple logic of forming request string into two

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 3 above)
Comment (pavertomato) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: it does, logic of this join quite straight forward


    public static String join(Collection<String> collection, String joinString) {
        StringBuilder builder = new StringBuilder();
        for (String str : collection) {
            if (builder.length() > 0) {
                builder.append(joinString);
            }
            builder.append(str);
        }
        return builder.toString();
    }

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 2 above)
Comment (poojavenkatram) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: Yes but the request string logic is anyway split intp two in the if and else cases...

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 3 above)
Comment (poojavenkatram) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: cool thanks.

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 1 above)
Comment (pavertomato) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: is it should be checkbox near patient name field or something else?
image

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 2 above)
Comment (pavertomato) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: well, this if-else populate list and person who reads may skip this part without losing sense of method
but anyway if you still think two return statement is better i'll revert this changes in one moment

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 1 above)
Comment (poojavenkatram) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: Yup that makes sense.

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 2 above)
Comment (poojavenkatram) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: Yeah it makes readability much better than the if-else blocks.

Comment (pavertomato) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: done

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 1 above)
Comment (pavertomato) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: image
it looked weird, so i changed it a bit
image

Comment (poojavenkatram) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: Looks good thanks :)

File: src/main/java/org/weasis/dicom/google/api/ui/SearchPanel.java [hunk 4]
```diff
@@ -43,10 +44,12 @@
 
 import static javax.swing.BoxLayout.LINE_AXIS;
 import static javax.swing.BoxLayout.PAGE_AXIS;
+import static javax.swing.BoxLayout.X_AXIS;
 
 public class SearchPanel extends JPanel {
 
     private final JTextField patientName = textField();
+    private final JCheckBox fuzzyMatching = textBox("fuzzy match", true);
```
Comment (poojavenkatram) [inline src/main/java/org/weasis/dicom/google/api/ui/SearchPanel.java]: The fuzzy match checkbox should be false by default.

File: src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java (hunk 2 above)
Comment (poojavenkatram) [reply src/main/java/org/weasis/dicom/google/api/GoogleAPIClient.java]: Thanks :)

File: src/main/java/org/weasis/dicom/google/api/ui/SearchPanel.java (hunk 4 above)
Comment (pavertomato) [reply src/main/java/org/weasis/dicom/google/api/ui/SearchPanel.java]: done

**Comment by pavertomato** [comment]
> Thanks :) Almost there :) Can we also add a Unit Test for checking that the Query is formed correctly with the new Param?

i've added this test, 
but this project not covered with test at all,
also seems like it need a lot of refactor,
so i think it worth to create separated tasks and prioritize this tasks  to achieve at least minimal test coverage and good looking code base

**Comment by poojavenkatram** [comment]
> > Thanks :) Almost there :) Can we also add a Unit Test for checking that the Query is formed correctly with the new Param?
> 
> i've added this test,
> but this project not covered with test at all,
> also seems like it need a lot of refactor,
> so i think it worth to create separated tasks and prioritize this tasks to achieve at least minimal test coverage and good looking code base

Thanks for adding this test - yeah completely agree we should have tasks to go back and add tests for the code already present. But going forward we should just start having Unit tests for all the new features so that we don't accumulate more code that isn't unit tested.

**Comment by pavertomato** [comment]
> > > Thanks :) Almost there :) Can we also add a Unit Test for checking that the Query is formed correctly with the new Param?
> > 
> > 
> > i've added this test,
> > but this project not covered with test at all,
> > also seems like it need a lot of refactor,
> > so i think it worth to create separated tasks and prioritize this tasks to achieve at least minimal test coverage and good looking code base
> 
> But going forward we should just start having Unit tests for all the new features 
i didn't argued with that

**Pull request merged**